Unquote paths

Cox65 opened the pull request:
It seems that there is an issue with path parameters which are not decoded before being passed to the WSGI application.
Looks like it is a common discussion when implementing WSGI adapter: https://github.com/benoitc/gunicorn/pull/1211

In my case I was receiving my path parameter still encoded in my Flask application.

Based on my researches, here is the fix.

adamchainz reviewed (changes requested):
Thanks for this! It looks like a rare gnarly issue.

Have you got an example path that you could add into some tests to prevent regressions?

Also, wouldn't `unquote(_, encoding="iso-8859-1")` work just as well, and avoid the trip through bytes and back?

Cox65 commented:
I made changes and added tests.
I introduced an optional parameter to the make_lambda_handler to be able to disable path unquoting as it seems that for some WSGI app it is not necessary. This way it can manage both behaviours.

adamchainz commented:
> I introduced an optional parameter to the make_lambda_handler to be able to disable path unquoting as it seems that for some WSGI app it is not necessary. 

When you say "not necessary", you mean it isn't strictly required because no URL path needs it? If so, I wouldn't want to make this a switch. Other users won't know what this means. We should just do the right thing every time. For example, gunicorn always decodes the path correctly...

Cox65 commented:
As I was not able to find a strict answer so I prefered to add this option but you are right, in this case, we need to follow PEP 3333 specifications.
As described here : https://github.com/benoitc/gunicorn/pull/1211#issuecomment-188258878
I remove this part and commit.

adamchainz commented:
Thanks! Released in version 2.9.2.
